File: 872.Leaf-SimilarTrees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    function getLeavesDFS(node) {
        let leaves = '';
        const stack = [];
        let current = node;
        while(true) {
            if (current) {
                stack.push(current);
                current = current.left;
            } else if (stack.length) {
                const node = stack.pop();
                if (!node.left && !node.right) {
                    leaves += ',' + node.val; // In case it's number string, we split the number with comma
                }
                current = node.right;
            } else {
                break;
            }
        }
        return leaves;
    }
    return getLeavesDFS(root1) === getLeavesDFS(root2);
};

File: Golang/Tree/872.Leaf-similar-trees.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	getLeafNodes := func(root *TreeNode) []int {
		ret := []int{}
		current := root
		stack := []*TreeNode{}
		for {
			if current != nil {
				stack = append(stack, current)
				current = current.Left
			} else if len(stack) > 0 {
				node := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				if node.Left == nil && node.Right == nil {
					ret = append(ret, node.Val)
				}
				current = node.Right
			} else {
				break
			}
		}
		return ret
	}
	root1Leaves := getLeafNodes(root1)
	root2Leaves := getLeafNodes(root2)
	if len(root1Leaves) == len(root2Leaves) {
		for i := 0; i < len(root1Leaves); i++ {
			if root1Leaves[i] != root2Leaves[i] {
				return false
			}
		}
		return true
	}
	return false
}
